pool/opaque.go

```diff
@@ -0,0 +1,97 @@
+package pool
+
+import (
+	"reflect"
+	"unsafe"
+)
+
+// OpaquePoolOptions specify options for OpaquePool.
+type OpaquePoolOptions struct {
+	Construct func(ptr interface{})
+	Size      int
+	Type      reflect.Type
+}
+
+// OpaquePool represents an object pool which is opaque to the GC.
+type OpaquePool struct {
+	heap []uint8
+	free chan uintptr
+	opts OpaquePoolOptions
+	step uintptr
+}
+
+// NewOpaquePool constructs a new OpaquePool.
+func NewOpaquePool(opts OpaquePoolOptions) *OpaquePool {
+	p := &OpaquePool{
+		free: make(chan uintptr, opts.Size),
+		opts: opts,
+		// TODO(@kobolog): alignment & padding.
+		step: hsz + opts.Type.Size(),
+	}
+
+	if p.opts.Construct == nil {
+		p.opts.Construct = func(interface{}) {}
+	}
+
+	p.init()
+
+	return p
+}
+
+type hdr struct {
+	// Offset from the beginning of the heap to the object.
+	offset uintptr
+}
+
+const (
+	hsz = unsafe.Sizeof(*(*hdr)(nil))
+)
+
+func (p *OpaquePool) init() {
+	// Heap is a slice of uint8 large enough to fit opts.Size objects
+	// of type struct { hdr; T }.
+	p.heap = make([]byte, p.opts.Size*int(p.step))
+
+	for i := uintptr(0); i < uintptr(len(p.heap)); i += p.step {
+		hdr := (*hdr)(unsafe.Pointer(&p.heap[i]))
+		hdr.offset = i + hsz
+		ptr := unsafe.Pointer(&p.heap[hdr.offset])
+
+		p.opts.Construct(reflect.NewAt(p.opts.Type, ptr).Interface())
+
+		// Alternatively, we can use a channel of interfaces, but it
+		// would keep opts.Size objects rooted in the channel.
+		// We aim to avoid allocating unnecessary GC-visible objects.
+		p.free <- hdr.offset
+	}
+}
+
+// Get provides an object from the pool.
+func (p *OpaquePool) Get() interface{} {
+	var val interface{}
+
+	select {
+	case offset := <-p.free:
+		val = reflect.NewAt(
+			p.opts.Type, unsafe.Pointer(&p.heap[offset])).Interface()
+	default:
+		val = reflect.New(p.opts.Type).Interface()
+		p.opts.Construct(val)
+	}
+
+	return val
+}
+
+// Put returns an object to the pool.
+func (p *OpaquePool) Put(object interface{}) {
+	ptr := unsafe.Pointer(reflect.ValueOf(object).Pointer())
+
+	if uintptr(ptr) < uintptr(unsafe.Pointer(&p.heap[0])) ||
+		uintptr(ptr) > uintptr(unsafe.Pointer(&p.heap[len(p.heap)-1])) {
+		return
+	}
+
+	// We know that this object is on our heap, so it's okay
+	// to read memory directly in front of it.
+	p.free <- (*hdr)(unsafe.Pointer(uintptr(ptr) - hsz)).offset
```

Comment:
Maybe we should check the address and see if it was from the pool, and if so directly enqueue it like we are here, but otherwise we just need to let the GC collect it.  i.e. we can't put normal heap allocated object address back in the pool.